### PR TITLE
core.sys.elf: Move shared ELF types and values to a common module

### PR DESCRIPTION
**Pull request by ibuclaw**
As noted in #3780, core.sys.linux.elf is really bindings for a Glibc header, rather than a Linux one.  However, the same structures and constants crop up across all OS-specific modules, of which the BSD ones are all copies of the FreeBSD elf module.

This commonizes it all into a core.sys.elf package.  Per-OS macros and values still remain.

Bootstrapped on x86_64-pc-solaris2.11 without any issues, lets see what the testsuite has to say about Linux and FreeBSD.

### CODE REVIEW

File: src/core/sys/solaris/sys/elf.d [hunk 1]
```diff
@@ -363,165 +54,28 @@ enum SHT_SUNW_SIGNATURE = 0x6ffffff6;
...
-enum STT_LOOS    = 10;
-enum STT_HIOS    = 12;
-enum STT_LOPROC  = 13;
-enum STT_HIPROC  = 15;
 
 extern (D)
 {
     auto ELF32_ST_VISIBILITY(O)(O o) { return o & 0x07; }
     alias ELF32_ST_VISIBILITY ELF64_ST_VISIBILITY;
 }
```
Comment (ibuclaw) [inline src/core/sys/solaris/sys/elf.d:65]: For reviewers:
Solaris defines `ELF32_ST_VISIBILITY` differently to all other Linux/BSDs (where it's `return o & 0x03`).

File: src/core/sys/solaris/sys/elf.d [hunk 2]
```diff
@@ -363,165 +54,28 @@ enum SHT_SUNW_SIGNATURE = 0x6ffffff6;
...
     auto ELF64_R_TYPE_DATA(I)(I i) { return (i << 32) >> 40; }
     auto ELF64_R_TYPE_ID(I)(I i) { return (i << 56) >> 56; }
     auto ELF64_R_TYPE_INFO(S, T)(S sym, T type) { return (sym <<8) + (type); }
 }
 
-enum GRP_COMDAT = 0x01;
-
 struct Elf32_Nhdr
 {
     Elf32_Word n_namesz;
```
Comment (ibuclaw) [inline src/core/sys/solaris/sys/elf.d:81]: For reviewers:
Regarding ELF notes, while there might be some common note sections, these are largely platform-specific.

File: src/core/sys/solaris/sys/elf.d [hunk 3]
```diff
@@ -536,35 +90,6 @@ struct Elf64_Nhdr
...
-    Elf64_Half  m_stride;
-}
-
-alias ELF32_M_SYM ELF64_M_SYM;
-alias ELF32_M_SIZE ELF64_M_SIZE;
-alias ELF32_M_INFO ELF64_M_INFO;
-
 struct Elf32_Cap
 {
     Elf32_Word  c_tag;
```
Comment (ibuclaw) [inline src/core/sys/solaris/sys/elf.d:95]: For reviewers:
The ELF capability table is a platform specific section.

**Review by dkorpel — COMMENTED** [review]
@WalterBright has been against efforts to factor out common definitions between platforms because of the following pattern:
- Even though there's a lot in common, there's still a lot of subtle differences sprinkled all over the place
- Someone encounters an error in the definition on one platform
- The definition is located and fixed
- Oops! Because the versioning logic wasn't simple, other platforms were inadvertently changed as well, and now they're broken

Now if it really is a Glibc header and the separation between the common module and platform-specific parts is rock solid, then I think we're okay. I'm don't have much experience with different linux platforms, but I trust your judgement.

**Comment by ibuclaw** [comment]
> @WalterBright has been against efforts to factor out common definitions between platforms because of the following pattern:
> - Even though there's a lot in common, there's still a lot of subtle differences sprinkled all over the place
> - Someone encounters an error in the definition on one platform
> - The definition is located and fixed
> - Oops! Because the versioning logic wasn't simple, other platforms were inadvertently changed as well, and now they're broken
> 

There's also libelf for non-elf platforms, so to say these are OS specific is wrong.

> Now if it really is a Glibc header and the separation between the common module and platform-specific parts is rock solid, then I think we're okay. I'm don't have much experience with different linux platforms, but I trust your judgement.

It says glibc/elf/elf.h in the reference URL, not glibc/sysdeps/linux/elf/elf.h.

So any platform that uses glibc as its primary C runtime would have the same bindings - Linux, Hurd, kFreebsd, kSolaris. Do we really want 3000 lines of bindings copied 4 times?

Also wrt Linux, has anyone consulted uclibc, musl? Maybe there's differences between them and glibc, in which case all of core.sys.linux.elf is wrong.

**Review by dkorpel — APPROVED** [review]
I suggest waiting a bit before merging this so @WalterBright (or someone else) has a chance to chime in.

**Comment by ibuclaw** [comment]
> I suggest waiting a bit before merging this so @WalterBright (or someone else) has a chance to chime in.

Having a talk with some people, there is the gABI spec (https://refspecs.linuxfoundation.org/elf/gabi4+/contents.html), and well as many psABI's to describe ELF for a specific processor (x86_64, SPARC, IA64, RISCV). It'll take time to unpick, but what's described there could be used as the baseline, and everything else thrown in os bindings.

It probably won't reduce the amount of duplication by 4000, but it will by some measure though.

**Comment by dkorpel** [comment]
@ibuclaw I think this can be merged now, or do you want to supersede this and go straight to bindings based on the gABI spec?

**Comment by dkorpel** [comment]
Ping @ibuclaw

**Pull request merged**